azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/PartitionManager.java

```diff
@@ -40,43 +43,68 @@
         this.host = host;
     }
     
-    String[] getPartitionIds() throws IllegalEntityException
+    CompletableFuture<String[]> getPartitionIds()
     {
-        Throwable saved = null;
-
-        if (this.partitionIds == null)
+    	CompletableFuture<String[]> retval = null;
+    	
+    	if (this.partitionIds != null)
+    	{
+    		retval = CompletableFuture.completedFuture(this.partitionIds);
+    	}
+    	else
         {
-			try
-			{
-				EventHubClient ehClient = EventHubClient.createFromConnectionStringSync(this.host.getEventHubConnectionString());
-				EventHubRuntimeInformation ehInfo = ehClient.getRuntimeInformation().get();
-				if (ehInfo != null)
+    		// This try-catch is necessary because EventHubClient.createFromConnectionString can directly throw
+    		// EventHubException or IOException, in addition to whatever failures may occur when the result of
+    		// the CompletableFuture is evaluated.
+    		try
+    		{
+    			// Stage 0: get EventHubClient for the event hub
+				retval = EventHubClient.createFromConnectionString(this.host.getEventHubConnectionString())
```

Comment:
>) [](start = 94, length = 1)

i think it will come - once you merge - but, just calling it out so its not missed..
- pass the executor down